neuralpp/symbolic/z3_expression.py

```diff
@@ -0,0 +1,335 @@
+            # So we have to create a function application in z3 and retrieve its declaration using decl().
+            return z3.And(True, True).decl()
+        case operator.or_:
+            return z3.Or(True, True).decl()
+        case operator.not_:
+            return z3.Not(True).decl()
+        case operator.xor:
+            return z3.Xor(True, True).decl()
+
+    x, y = z3.Consts("x y", type_to_z3_sort(type_) if type_ is not None else z3.IntSort())
+    match python_callable:
+        # comparison and arithmetic are overloaded by z3.
+        case operator.le:
+            return (x <= y).decl()
+        case operator.lt:
+            return (x < y).decl()
+        case operator.ge:
+            return (x >= y).decl()
+        case operator.gt:
+            return (x > y).decl()
+        case operator.eq:
+            return (x == y).decl()
+        case operator.add:
+            return (x + y).decl()
+        case operator.mul:
+            return (x * y).decl()
+        case operator.pow:
+            return (x ** y).decl()
+        # min/max
+        case builtins.min:
+            raise NotImplementedError("Cannot convert min to a z3 function declaration."
+                                      "However we can create z3.If(x<y, x, y) for min(x,y).")
+        case builtins.max:
+            # if len(arguments) != 2:
+            #     raise NotImplementedError("Only 2-element max is supported")
+            # return z3.If(arguments[0] > arguments[1], arguments[0], arguments[1])
+            raise NotImplementedError("Cannot convert min to a z3 function declaration."
+                                      "However we can create z3.If(x>y, x, y) for max(x,y).")
+        case _:
+            raise ValueError(f"Python callable {python_callable} is not recognized.")
+
+
+def z3_function_to_python_callable(z3_function: z3.FuncDeclRef) -> Callable:
+    match z3_function.kind():
+        # boolean operation
+        case z3.Z3_OP_AND:
+            return operator.and_
+        case z3.Z3_OP_OR:
+            return operator.or_
+        case z3.Z3_OP_NOT:
+            return operator.not_
+        case z3.Z3_OP_XOR:
+            return operator.xor
+        # comparison
+        case z3.Z3_OP_LE:
+            return operator.le
+        case z3.Z3_OP_LT:
+            return operator.lt
+        case z3.Z3_OP_GE:
+            return operator.ge
+        case z3.Z3_OP_GT:
+            return operator.gt
+        case z3.Z3_OP_EQ:
+            return operator.eq
+        # arithmetic
+        case z3.Z3_OP_ADD:
+            return operator.add
+        case z3.Z3_OP_MUL:
+            return operator.mul
+        case z3.Z3_OP_POWER:
+            return operator.pow
+        case _:
+            raise ValueError(f"Z3 function {z3_function} is not recognized.")
+
+
+def z3_object_to_expression(z3_object: z3.ExprRef) -> Z3Expression:
+    # Here we just try to find a type of expression for z3 object.
```

Comment:
I think this comment can be a bit more precise: you are not just trying to find a type, but to build the corresponding `Expression` object. How about "Returns an instance of the appropriate subclass of `Expression` matching `z3_object`"?